qa/rpc-tests/invalidateblock.py

```diff
@@ -89,6 +89,43 @@ def run_test(self):
 
         self.testChainSyncWithLongerInvalid()
 
+        self.testMempoolDuringInvalidate()
+
+    def testMempoolDuringInvalidate(self):
+        """ This test checks dependent transactions committed and in the mempool during block invalidates and verifies that the dependent one does not
+            stay in the mempool if the dependency is removed.
+        """
+        # start this test connected
+        connect_nodes_bi(self.nodes,1,2)
+        # Get something to spend
+        nBlocks = self.nodes[0].getblockcount()
+        self.nodes[0].generate(101-nBlocks)
+        nBlocks = 101
+
+        sync_blocks(self.nodes)
+
+        # Set up 2 dependent tx, 1 committed the other in mempool.  invalidate the committed block and make sure that we don't end up
+        # with tx 2 alone in the mempool.
+        addr = self.nodes[1].getnewaddress()
+        tx1hash = self.nodes[0].sendtoaddress(addr, 20)
+        waitFor(10, lambda: self.nodes[0].getmempoolinfo()["size"] > 0)
+        block = self.nodes[0].generate(1)[0]
+        waitFor(5, lambda: self.nodes[1].getblockcount() == nBlocks+1)
+        tx2hash = self.nodes[1].sendtoaddress(addr, 10)  # This must be dependent on the prior send because node 1 has no other money
+        waitFor(10, lambda: self.nodes[0].getmempoolinfo()["size"] == 1)
+        self.nodes[0].invalidateblock(block)
+        mp = self.nodes[0].getrawmempool()
+        assert(tx1hash in mp)
+        # tx2 won't be in the mempool because of its nLockTime (see fee sniping)
```

Comment:
This comment assertion could be a code assertion :-)

Also this is not true, 10% chance nLockTime is != tip height. See wallet/wallet.cpp:2278

Reply:
This needs to be fixed so we don't get random test failures, but the rest of the PR LGTM

Reply:
the 10% chance is why I can't assert... I'll fix the comment, but since I'm not checking whether it got in or now, how would I get test failures?

Reply:
You won't, my bad